game-core/src/main/java/games/strategy/engine/framework/startup/ui/InGameLobbyWatcher.java

```diff
@@ -277,21 +264,17 @@ private void cleanUpGameModelListener() {
   }
 
   private void updatePlayerCount() {
-    synchronized (mutex) {
-      gameDescription.setPlayerCount(serverMessenger.getNodes().size() - (isHandlerPlayer ? 0 : 1));
-      postUpdate();
-    }
+    postUpdate(gameDescription.get().withPlayerCount(serverMessenger.getNodes().size() - (isHandlerPlayer ? 0 : 1)));
   }
 
-  private void postUpdate() {
-    if (isShutdown) {
+  private void postUpdate(final GameDescription newDescription) {
+    if (isShutdown || newDescription.equals(gameDescription.get())) {
```

Comment:
I'll have to revisit this code anyways. I realised that the atomic reference doesn't solve any problems and just adds complexity.
Because the object is now immutable we might end up creating 2 different objects at the same time and only pushing a single one because the other one just lost the race for the lock.
So at some point a mutex is required to synchronise changes to the GameProperties object.

Reply:
Okay, though I'd caution that any race conditions are pretty unlikely. We need consistency at a user-facing level. If for example we spend a few milliseconds waiting for a network request, solving a race condition of a few nano-seconds is not going to matter. The user could get a TCP retry in there and get the wrong answer anyways. I'm not saying ignore locking, but we need to be sure we have the right requirements in mind.